OWML.ModLoader/ModSorter.cs

```diff
@@ -0,0 +1,122 @@
+            var modList = new List<string>();
+            foreach (var mod in mods)
+            {
+                 modList.Add(mod.Manifest.UniqueName);
+            }
+
+            // Make hashset of tuples of Dependant : Dependency
+            HashSet<Tuple<string, string>> set = new HashSet<Tuple<string, string>>();
+            foreach (var mod in mods)
+            {
+                foreach (var dep in mod.Manifest.Dependencies)
+                {
+                    set.Add(new Tuple<string, string>(mod.Manifest.UniqueName, dep));
+                }
+            }
+
+            // Sort the mods
+            var ret = TopologicalSort(
+                new HashSet<string>(modList),
+                new HashSet<Tuple<string, string>>(set)
+            );
+
+            // Reverse the list
+            ret.Reverse();
+
+            // Get the IModData back by looking up uniqueName in dict
+            var returnList = new List<IModData>();
+            ret.ForEach(x => returnList.Add(modDict[x]));
+
+            return returnList;
+        }
+
+        // Thanks to https://gist.github.com/Sup3rc4l1fr4g1l1571c3xp14l1d0c10u5/3341dba6a53d7171fe3397d13d00ee3f
+
+        static List<T> TopologicalSort<T>(HashSet<T> nodes, HashSet<Tuple<T, T>> edges)
+        {
+            // Empty list that will contain the sorted elements
+            var sortedList = new List<T>();
+
+            // Set of all nodes with no incoming edges
+            var nodesWithNoEdges = new HashSet<T>(nodes.Where(n => edges.All(e => e.Item2.Equals(n) == false)));
+
+            // while nodesWithNoEdges is non-empty do
+            while (nodesWithNoEdges.Any())
+            {
+                //  remove a node from nodesWithNoEdges
+                var node = nodesWithNoEdges.First();
+                nodesWithNoEdges.Remove(node);
+
+                // add node to tail of sortedList
+                sortedList.Add(node);
+
+                // for each node m with an edge e from n to m do
+                foreach (var e in edges.Where(e => e.Item1.Equals(node)).ToList())
+                {
+                    var m = e.Item2;
+
+                    // remove edge e from the graph
+                    edges.Remove(e);
+
+                    // if m has no other incoming edges then
+                    if (edges.All(me => me.Item2.Equals(m) == false))
+                    {
+                        // insert m into nodesWithNoEdges
+                        nodesWithNoEdges.Add(m);
+                    }
+                }
+            }
+
+            // if graph has edges then
+            if (edges.Any())
+            {
+                // return error (graph has at least one cycle)
+                return null;
+            }
+            else
+            {
+                // return sortedList (a topologically sorted order)
+                return sortedList;
+            }
+        }
+    }
+
+    public class Tuple<T1, T2>
```

Comment:
Do we only use tuples for graph edges? If so, why not call this class `Edge`?
And, if tuples are only used for edges, they also only need to work with nodes (which are actually just strings?). So they don't need to be generic. You can just define the class Edge as having attributes `string From` and `string To`, instead of the more generic `Item1` and `Item2`.

In other situations it may make sense to have something super generic to work with everything, but in our case we know that we only want to use this sort for sorting mods, so we can type and name everything accordingly. This will make the code a lot more readable and reduce the type parameter noise.